src/java/com/puppetlabs/jruby_utils/pool/ReferencePool.java

```diff
@@ -0,0 +1,526 @@
+        // `LinkedBlockingDeque` does for a timed poll from a deque.  See:
+        // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/LinkedBlockingDeque.java#l516
+        lock.lockInterruptibly();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            // This pattern of using timed `awaitNanos` on a condition
+            // variable to track the total time spent waiting for an item to
+            // be available to be borrowed follows the logic that the JDK's
+            // `LinkedBlockingDeque` in `pollFirst` uses.  See:
+            // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/LinkedBlockingDeque.java#l522
+            do {
+                if (this.pill != null) {
+                    // Return the pill immediately if there is one
+                    item = pill;
+                } else if (isPoolLockHeldByAnotherThread(currentThread)) {
+                    if (remainingMaxTimeToWait <= 0) {
+                        break;
+                    }
+                    remainingMaxTimeToWait =
+                            poolNotLocked.awaitNanos(remainingMaxTimeToWait);
+                } else if (this.instance == null) {
+                    // No instance initialized yet
+                    if (remainingMaxTimeToWait <= 0) {
+                        break;
+                    }
+                    remainingMaxTimeToWait =
+                            borrowsAvailable.awaitNanos(remainingMaxTimeToWait);
+                } else if (this.currentBorrowCount.get() >= this.maxBorrowCount) {
+                    // Max borrow count reached, wait for one to be returned
+                    if (remainingMaxTimeToWait <= 0) {
+                        break;
+                    }
+                    remainingMaxTimeToWait =
+                            borrowsAvailable.awaitNanos(remainingMaxTimeToWait);
+                } else if (instance != null) {
+                    item = instance;
+                    this.currentBorrowCount.getAndIncrement();
+                    LOGGER.info("Borrowing, count is " + currentBorrowCount.get());
+                }
+            } while (item == null);
+        } finally {
+            lock.unlock();
+        }
+
+        return item;
+    }
+
+    /**
+     * Release an item and return it to the pool. Does nothing if the item
+     * being released is the pill.
+     * Throws an `IllegalArgumentException` if the item is not currently
+     * registered by the pool and the item is not the pill, if one has been
+     * inserted
+     */
+    @Override
+    public void releaseItem(E e) {
+        final ReentrantLock lock = this.borrowLock;
+        lock.lock();
+        try {
+            if (e != this.pill) {
+                if (!isRegistered(e)) {
+                    String errorMsg = "The item being released is not registered with the pool";
```

Comment:
Would it be useful to include what the item is in the error message?

Reply:
What sort of information do you think would be useful here? It's kind of ugly to print an object ID, maybe just adding the type information? Not sure if this could get ugly if it's a type coming in from Clojure.

Reply:
I'm not sure, since I don't really know what object info is available or what a user might find actionable. Mostly just thinking generally/hypothetically about error messages and imagining myself seeing that message and then wondering "what item??" but I don't really know when/if that might come up.